pkg/crd/desc_visitor.go

```diff
@@ -0,0 +1,72 @@
+/*
+Copyright 2019 The Kubernetes Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package crd
+
+import (
+	"strings"
+
+	apiext "k8s.io/apiextensions-apiserver/pkg/apis/apiextensions/v1beta1"
+)
+
+// TruncateDescription truncates the description of fields in given schema if it
+// exceeds maxLen.
+// It tries to chop off the description at the closest sentence boundary.
+func TruncateDescription(schema *apiext.JSONSchemaProps, maxLen int) *apiext.JSONSchemaProps {
+	outSchema := schema.DeepCopy()
+	EditSchema(outSchema, descVisitor{maxLen: maxLen})
+	return outSchema
+}
+
+// descVisitor recursively visits all fields in the schema and truncates the
+// description of the fields to specified maxLen.
+type descVisitor struct {
+	// maxLen is the maximum allowed length for decription of a field
+	// 0 indicates unlimited length
+	// -1 indicates no description is allowed
+	maxLen int
+}
+
+func (v descVisitor) Visit(schema *apiext.JSONSchemaProps) SchemaVisitor {
+	if schema == nil {
+		return v
+	}
+	if v.maxLen == 0 {
+		return nil /* no further work to be done for this schema */
+	}
+	if v.maxLen < 0 {
+		schema.Description = ""
+		return v
+	}
+	if len(schema.Description) > v.maxLen {
+		schema.Description = truncateString(schema.Description, v.maxLen)
+		return v
+	}
+	return v
+}
+
+func truncateString(desc string, maxLen int) string {
+	desc = desc[0:maxLen]
+
+	// Trying to chop off at closest sentence boundary.
+	if n := strings.LastIndexAny(desc, "."); n > 0 {
```

Comment:
". " (period-space) is safer for sentences, but we should actually probably use unicode.SentenceTerminal or something similar.

Reply:
I couldn't find some quick helper for SentenceTerminal in `unicode` pkg --> https://golang.org/pkg/unicode/. Quick search suggest, I can create a helper fn using chars listed in this page --> https://www.unicode.org/review/pr-23.html. Do you think it is worth the effort here ?